src/lib/zs_matcher.rs

```diff
@@ -70,36 +147,361 @@ Fast Algorithms for the Editing Distance Between Trees and Related Problems.";
         String::from(desc)
     }
 
-    /// Match locations in distinct ASTs.
-    fn match_trees(&self, base: Arena<T, FromNodeId>, diff: Arena<T, ToNodeId>) -> MappingStore<T> {
-        MappingStore::new(base, diff)
+    /// Match locations in distinct ASTs (not described in the paper).
+    fn match_trees(&mut self, base: Arena<T, FromNodeId>, diff: Arena<T, ToNodeId>)
+                   -> MappingStore<T> {
+        let store = MappingStore::new(base, diff);
+        self.compute_tree_distance(&store);
+        let mut root_pair = true;
+        let mut tree_pairs: Vec<(usize, usize)> = vec![];
+        tree_pairs.push((self.src.node_count, self.dst.node_count));
+        while !tree_pairs.is_empty() {
+            let tree_pair = tree_pairs.pop().unwrap();
+            let last_row = tree_pair.0;
+            let last_col = tree_pair.1;
+            if !root_pair {
+                self.forest_dist(last_row, last_col, &store);
+            }
+            root_pair = false;
+            let first_row = self.src.lld(last_row) - 1;
+            let first_col = self.dst.lld(last_col) - 1;
+            let mut row = last_row;
+            let mut col = last_col;
+            while row > first_row || col > first_col {
+                if row > first_row
+                   && self.forest_dist[row - 1][col] + 1.0 == self.forest_dist[row][col]
```

Comment:
I get nervous about any floating point calculation that relies on `==` or `!=` -- are we sure this is doing what we think it is?

Reply:
IIWYM, although this is taken quite directly from the GT code. Perhaps an epsilon would be better here?

Reply:
IIWYM means ...?

We can't blindly add an epsilon without thinking about what's going on -- unfortunately there isn't going to be a substitute for thinking about what this is really doing. For example, how many decimal places might we be caring about? could/should we be converting this to integer arithmetic? and so on.

Reply:
If we use floats as the representation, then not even all integers are accurately represented. I don't think we can merge this until we have some answers to the questions I suggested earlier. I know this sucks, but if no-one else has been bothered thinking about this in the past, we might have to do so now...